File: leetcode/meta/1249.cpp
#include <string>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<int> stk;
        string ans;
        unordered_set<int> to_delete;
        for (int i = 0; i < s.size(); ++i) {
            if (stk.empty() && s[i] == ')') {
                to_delete.insert(i);
            } else if (s[i] == '(') {
                stk.push_back(i);
            } else if (s[i] == ')') {
                stk.pop_back();
            }
        }
        for (int index : stk) {
            to_delete.insert(index);
        }
        for (int i = 0; i < s.size(); ++i) {
            if (!to_delete.count(i)) {
                ans.push_back(s[i]);
            }
        }
        return ans;
    }
};

File: leetcode/1249.py
from typing import *

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        stack = []
        to_delete = set()
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            elif s[i] == ')':
                if len(stack) == 0:
                    to_delete.add(i)
                else:
                    stack.pop()
        to_delete |= set(stack)
        ans = ""
        for i in range(len(s)):
            if i not in to_delete:
                ans += s[i]
        return ans